File: core/src/main/java/y2k/joyreactor/images/DiskCache.java
package y2k.joyreactor.images;

import rx.Observable;
import rx.schedulers.Schedulers;
import y2k.joyreactor.common.Optional;
import y2k.joyreactor.platform.Platform;

import java.io.File;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

/**
 * Created by y2k on 9/27/15.
 */
class DiskCache {

    private static Executor DISK_EXECUTOR = Executors.newSingleThreadExecutor();

    public DiskCache() {
        getCacheDirectory().mkdirs();
    }

    Observable<Optional<File>> loadOptionalAsync(String url) {
        return Observable.<Optional<File>>create(subscriber -> {
            File file = urlToFile(url);
            subscriber.onNext(file.exists() ? Optional.of(file) : Optional.<File>empty());
            subscriber.onCompleted();
        }).subscribeOn(Schedulers.from(DISK_EXECUTOR));
    }

    Observable<?> putAsync(File newImageFile, String url) {
        return Observable.create(subscriber -> {
            newImageFile.renameTo(urlToFile(url));
            subscriber.onNext(null);
            subscriber.onCompleted();
        }).subscribeOn(Schedulers.from(DISK_EXECUTOR));
    }

    private File urlToFile(String url) {
        return new File(getCacheDirectory(), "" + url.hashCode());
    }

    File getCacheDirectory() {
        return new File(Platform.Instance.getCurrentDirectory(), "images");
    }
}

File: core/src/main/kotlin/y2k/joyreactor/images/DiskCache.kt
package y2k.joyreactor.images

import rx.Observable
import rx.schedulers.Schedulers
import y2k.joyreactor.platform.Platform
import java.io.File
import java.util.concurrent.Executors

/**
 * Created by y2k on 9/27/15.
 */
internal class DiskCache {

    init {
        cacheDirectory.mkdirs()
    }

    operator fun get(url: String): Observable<File?> {
        return Observable.create<File?> { subscriber ->
            val file = urlToFile(url)
            subscriber.onNext(if (file.exists()) file else null)
            subscriber.onCompleted()
        }.subscribeOn(Schedulers.from(DISK_EXECUTOR))
    }

    fun put(newImageFile: File, url: String): Observable<*> {
        return Observable.create<Any> { subscriber ->
            newImageFile.renameTo(urlToFile(url))
            subscriber.onNext(null)
            subscriber.onCompleted()
        }.subscribeOn(Schedulers.from(DISK_EXECUTOR))
    }

    private fun urlToFile(url: String): File {
        return File(cacheDirectory, "" + url.hashCode())
    }

    val cacheDirectory: File
        get() = File(Platform.Instance.currentDirectory, "images")

    companion object {

        private val DISK_EXECUTOR = Executors.newSingleThreadExecutor()
    }
}